epochStart/bootstrap/epochStartDataProvider.go

```diff
@@ -0,0 +1,512 @@
+		if epochStartData.ShardID == shardID {
+			return &epochStartData, nil
+		}
+	}
+
+	return nil, errors.New("not found")
+}
+
+func (esdp *epochStartDataProvider) initTopicsAndInterceptors() error {
+	err := esdp.messenger.CreateTopic(factory.MetachainBlocksTopic, true)
+	if err != nil {
+		log.Info("error unregistering message processor", "error", err)
+		return err
+	}
+
+	err = esdp.messenger.RegisterMessageProcessor(factory.MetachainBlocksTopic, esdp.epochStartMetaBlockInterceptor)
+	if err != nil {
+		return err
+	}
+
+	err = esdp.messenger.CreateTopic(factory.ShardBlocksTopic+"_1_META", true)
+	if err != nil {
+		log.Info("error unregistering message processor", "error", err)
+		return err
+	}
+
+	err = esdp.messenger.RegisterMessageProcessor(factory.ShardBlocksTopic+"_1_META", esdp.shardHeaderInterceptor)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (esdp *epochStartDataProvider) getShardID(nodesConfig *sharding.NodesSetup) (uint32, error) {
+	pubKeyBytes, err := esdp.publicKey.ToByteArray()
+	if err != nil {
+		return 0, err
+	}
+	pubKeyStr := hex.EncodeToString(pubKeyBytes)
+	for shardID, nodesPerShard := range nodesConfig.InitialNodesPubKeys() {
+		for _, nodePubKey := range nodesPerShard {
+			if nodePubKey == pubKeyStr {
+				return shardID, nil
+			}
+		}
+	}
+
+	return 0, nil
+}
+
+func (esdp *epochStartDataProvider) getTrieFromRootHash(_ []byte) (state.TriesHolder, error) {
+	// TODO: get trie from trie syncer
+	return state.NewDataTriesHolder(), nil
+}
+
+func (esdp *epochStartDataProvider) resetTopicsAndInterceptors() {
+	err := esdp.messenger.UnregisterAllMessageProcessors()
+	if err != nil {
+		log.Info("error unregistering message processors", "error", err)
+	}
+}
+
+func (esdp *epochStartDataProvider) getMetaBlock(hash []byte) (*block.MetaBlock, error) {
+	esdp.requestMetaBlock(hash)
+
+	time.Sleep(delayAfterRequesting)
+
+	for {
+		numConnectedPeers := len(esdp.messenger.Peers())
+		threshold := int(thresholdForConsideringMetaBlockCorrect * float64(numConnectedPeers))
+		mb, errConsensusNotReached := esdp.metaBlockInterceptor.GetMetaBlock(hash, threshold)
+		if errConsensusNotReached == nil {
+			return mb, nil
+		}
+		log.Info("consensus not reached for meta block. re-requesting and trying again...")
+		esdp.requestMetaBlock(hash)
+	}
+}
+
+func (esdp *epochStartDataProvider) getEpochStartMetaBlock(epoch uint32) (*block.MetaBlock, error) {
+	esdp.requestEpochStartMetaBlock(epoch)
+
+	time.Sleep(delayAfterRequesting)
+
+	for {
+		numConnectedPeers := len(esdp.messenger.Peers())
+		threshold := int(thresholdForConsideringMetaBlockCorrect * float64(numConnectedPeers))
+		mb, errConsensusNotReached := esdp.epochStartMetaBlockInterceptor.GetEpochStartMetaBlock(threshold, epoch)
+		if errConsensusNotReached == nil {
+			return mb, nil
+		}
+		log.Info("consensus not reached for meta block. re-requesting and trying again...")
+		esdp.requestEpochStartMetaBlock(epoch)
```

Comment:
there should be a max retry here as well. should not have infinit cylces. the same for all the other methods.

Reply:
this should be blocking, but I will add a limit and after the limit is exceeded, the app will panic